### DISCUSSION TITLE
Implementing SMAA (from pmndrs) into three js

### DISCUSSION THREAD

**Discussion by tabetVercetti** [pmndrs/postprocessing#557] [Q&A] [answered]
Hello all :D I am getting myself familiar with this wonderful postprocessing library. I went over the demo codes, but to be honest, i am struggling to understand how to implement some of these effects. I created a simple three js case in order to practice. So far i have managed to add selective bloom, but i would like to add anti-aliasing (SMAA). I would appreciate it if anyone can guide me on how i can add smaa into this simple three js case im working on: 
`//Imports for standard Threejs
import * as THREE from 'three';

//Imports for pmndrs post processing library
import { WebGLRenderer } from 'three';
import {
    SelectiveBloomEffect,
    BlendFunction,
    SMAAEffect,
    EffectComposer, 
    EffectPass, 
    RenderPass 
} from "postprocessing";

// Create scene
const scene = new THREE.Scene();

// Create red GLOWING cube BASIC
const redGeometry = new THREE.BoxGeometry();
const redMaterial = new THREE.MeshBasicMaterial({ color: new THREE.Color().setRGB(1, 0
  , 0) }); 
// color is set to 10 in order to go out of glow threshold 
const redCube = new THREE.Mesh(redGeometry, redMaterial);
redCube.position.x = -2; // Position to the left
redCube.material.toneMapped = false;
scene.add(redCube);
console.log(redCube)

// Create green cube (NON-GLOWING)
const greenGeometry = new THREE.BoxGeometry();
const greenMaterial = new THREE.MeshBasicMaterial({ color: 0x00ff00 });
const greenCube = new THREE.Mesh(greenGeometry, greenMaterial);
greenCube.position.x = 2; // Position to the right
scene.add(greenCube);

// Create camera
const camera = new THREE.PerspectiveCamera(
  75,
  window.innerWidth / window.innerHeight,
  0.1,
  1000
);
camera.position.z = 5;

// Create renderer
const renderer = new WebGLRenderer({
	powerPreference: "high-performance",
	antialias: false,
});

renderer.setSize(window.innerWidth, window.innerHeight);
document.body.appendChild(renderer.domElement);

//Create Effects
const bloomEffect = new SelectiveBloomEffect(scene, camera, {
  intensity: 5,
  mipmapBlur: true,
  luminanceThreshold: 0,
  luminanceSmoothing: 0.2,
  radius : 0.618,
  resolutionScale: 4
});

bloomEffect.selection.add(redCube);

const composer = new EffectComposer(renderer);
composer.addPass(new RenderPass(scene, camera));
composer.addPass(new EffectPass(camera, bloomEffect));

// Animation function
const animate = () => {
  requestAnimationFrame(animate);

  // Rotate cubes
  redCube.rotation.x += 0.01;
  redCube.rotation.y += 0.01;
  greenCube.rotation.x += 0.01;
  greenCube.rotation.y += 0.01;

  composer.render(scene, camera);
};

// Handle window resize
window.addEventListener('resize', () => {
  const newWidth = window.innerWidth;
  const newHeight = window.innerHeight;

  camera.aspect = newWidth / newHeight;
  camera.updateProjectionMatrix();

  renderer.setSize(newWidth, newHeight);
});

// Start animation
animate();
`

**Comment by vanruesc** [comment, marked as answer]
Hi, the [manual](https://github.com/pmndrs/postprocessing/tree/main/manual/assets/js/src/demos) (WIP) contains up-to-date examples that are a easier to follow. Also note that `composer.setSize` should be used instead of `renderer.setSize`.

**Comment by tabetVercetti** [comment]
Thanks @vanruesc for your helpful reply. I followed the manual demo and was able to implement smaa. I have a few questions though: 

Firstly, i noticed that the demo in the manual does not mention anything related to "area" and "search". These were present in other demos i saw. From what i could understand, they were textures. How come this demo does not include them? 

Also, I tried implementing SMAA with and without the debugging lines and noticed no significant difference. To be honest i didnt really understand what theyre doing but i am guessing its recommended that i also have them in my code? 

Lastly, one of the reasons why i was struggling at the beginning was because i was also wrapping my head around the load function and the loadmanager from three js. Previous demos required these in order for smaa to work but the demo in the manual doesnt. Just curious how come the methodology has changed. 

Thanks again for your support! Heres how its looking so far: 
[Codesandbox](https://codesandbox.io/p/sandbox/brave-tamas-y9g8xg?file=%2Fsrc%2FpmndrsBloomTest.js%3A56%2C38&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clp9u6b5100072a6cgxb6149h%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clp9u6b5100032a6cq3h2kxk1%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clp9u6b5100042a6c91b41d8y%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clp9u6b5100062a6cxxj0vv4e%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B40%252C60%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clp9u6b5100032a6cq3h2kxk1%2522%253A%257B%2522id%2522%253A%2522clp9u6b5100032a6cq3h2kxk1%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpazn6ho00032a6cxh9c2id4%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A56%252C%2522startColumn%2522%253A38%252C%2522endLineNumber%2522%253A56%252C%2522endColumn%2522%253A38%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FpmndrsBloomTest.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clpazn6ho00032a6cxh9c2id4%2522%257D%252C%2522clp9u6b5100062a6cxxj0vv4e%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clp9u6b5100052a6cbq1r6mba%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%2522%257D%255D%252C%2522id%2522%253A%2522clp9u6b5100062a6cxxj0vv4e%2522%252C%2522activeTabId%2522%253A%2522clp9u6b5100052a6cbq1r6mba%2522%257D%252C%2522clp9u6b5100042a6c91b41d8y%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clp9u6b5100042a6c91b41d8y%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Comment by vanruesc** [comment]
The search/area textures are data textures that can be loaded automatically by the `SMAAEffect` itself. The code was changed a while ago so that the user doesn't have to worry about it anymore.

The debugging code is not needed for SMAA. You only need the effect.

The `LoadingManager` is just a tool that helps with managing multiple parallel downloads. It fires a callback when all downloads have finished which is useful for loading scene assets before starting the render loop. The following sandbox uses it as well to provide a place where assets could be loaded if needed: https://codesandbox.io/p/sandbox/postprocessing-25rts?file=%2Fsrc%2FAssets.js%3A8%2C1

**Reply by tabetVercetti** [reply]
Once again thank you for your reply! its helping me learn more about this topic. 
I have also noticed that i can get anti aliasing through msaa simply by changing this line
 `const composer = new EffectComposer(renderer);` 
to this 
`const composer = new EffectComposer(renderer, {
    multisampling: Math.min(4, renderer.capabilities.maxSamples),
    frameBufferType: HalfFloatType
});` 
but i understand that this has limitations for devices that dont support webgl 2. Is it recommended that i have both msaa and smaa implemented in my project, or since i already implemented smaa, then theres no point in using msaa?

**Reply by vanruesc** [reply]
MSAA and SMAA both smooth jagged lines but they work very differently. SMAA is image-based and fast but suffers from shimmering. MSAA renders additional fragments along geometrical edges for stable smoothing but suffers from artifacting when used with depth-based effects like SSAO.

WebGL 1 is deprecated at this point so it's fine to just use WebGL 2. I recommend using MSAA as the default and SMAA for weaker devices or when MSAA produces artifacts.
